File: 354. Russian Doll Envelopes.java
class Solution {//Time of O(N^2+ N logN) and space O(N)
    public int maxEnvelopes(int[][] envelopes) {
        
        int len = envelopes.length;
        //Base case 
        if(len==0) return 0;
        
        Arrays.sort(envelopes, (a,b)-> Integer.compare(a[0],b[0]));//Compare width
        int[] dp = new int[len];
       // dp[0] = 1;
       Arrays.fill(dp,1);
        int result = 1;
        
        for(int i =1; i<len ; i++){
           // int maxlen = 0;
            for(int j=0 ; j<i ;j++){
                if(envelopes[i][0]> envelopes[j][0] && envelopes[i][1]>envelopes[j][1]){
                 //  maxlen = Math.max(maxlen,dp[j]);
                    dp[i] = Math.max(dp[i],dp[j]+1);
                }
                //dp[i] = 1 + maxlen;
                result = Math.max(result,dp[i]);
            }
        }
        return result;
    }
}

File: 354-RussianDollEnvelopes.cpp
/* 
    Time Complexity                              :  O(N*N) - dynamicProgramming
                                                    O(N Log N) - maxEnvelopesOptimized
    Space Complexity                             :  O(N) - to store the values in the dp array
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  


class Solution {
private:
    static bool cmp(const vector<int>& a, const vector<int>& b) {
        return a[0]==b[0] ? a[1] > b[1] : a[0] < b[0];
    }
public:
    int maxEnvelopes(vector<vector<int>>& envelopes) {
        return maxEnvelopesOptimized(envelopes);
        // return dynamicProgramming(envelopes);
    }
    // O(N*N) hence time limit exceeded
    int dynamicProgramming(vector<vector<int>>& env) {
        int n = env.size();
        sort(begin(env),end(env),cmp);
        vector<int> dp(n,1);
        int mx = 1;
        for(int i=1;i<n;i++) {
            for(int j=0; j<i;j++) {
                if(env[i][1] > env[j][1]) {
                    dp[i] = max(dp[i],dp[j] + 1);
                }
            }
            mx = max(mx, dp[i]);
        }
        
        return mx;
    }
    
    
    // O(N log N) since used binary search
    int maxEnvelopesOptimized(vector<vector<int>>& envelopes) {
        
        auto cmp = [](vector<int>& a, vector<int>& b) { 
            return a[0]==b[0] ? a[1] > b[1] : a[0] < b[0];
        };
        sort(begin(envelopes), end(envelopes), cmp);
        int n = envelopes.size();
        vector<int> dp;
        dp.push_back(envelopes[0][1]);
        for(int i=1;i<n;i++) {
            if(envelopes[i][1] > dp.back()) {
                dp.push_back(envelopes[i][1]);
            } else {
                int bsIndex = binarySearch(dp,0,dp.size()-1,envelopes[i][1]);
                // int bsIndex = lower_bound(dp.begin(),dp.end(),envelopes[i][1]) - dp.begin();
                dp[bsIndex] = envelopes[i][1];
            }
        }
        return dp.size();
    }
    
    
    int binarySearch(vector<int>& dp, int low, int high, int target) {
        while(low <= high) {
            int mid = low + (high - low)/2;
            if(dp[mid] == target) {
                return mid;
            } else if(dp[mid] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        
        return low;
    }
};